### PR TITLE
refactor: Service Event - Update to Register/Unregister APIs

### LINKED ISSUE #276: Service Events - Register/Unregister API update
Based on comments in https://github.com/hyperledger/aries-framework-go/pull/240#discussion_r325332114

The current code behaviour is as follows.

1. If consumer doesn't call RegisterEvent or RegisterAutoExecute - service will not handle messages and throw error for incoming messages.
2. If consumer doesn't call RegisterEvent and calls RegisterAutoExecute - service will auto execute the messages.
3. If consumer doesn't call RegisterAutoExecute and calls RegisterEvent - service will trigger the event and wait for callback
4. If RegisterAutoExecute and RegisterEvent both are called - service will behave same as pt#2 or pt#3 based on which was registered last.
5. If RegisterAutoExecute and RegisterEvent both are called AND service calls either UnregisterAutoExecute or  UnregisterEvent - service will behave same as pt#1.
6. If consumer calls RegisterEvent twice - last channel registered will receive the message.
7. If any consumer calls UnregisterEvent - service removes the event irrespective of the consumer

### PR DESCRIPTION
**Pull request by rolsonquadras**
- Remove AutoExecute Register/Unregister API
- Consumer need to register a channel to process all the messages
- Utility function to AutoExecute Action Events

Design is tracked in #276. This PR also closes #276.

Signed-off-by: Rolson Quadras <rolson.quadras@securekey.com>

### CODE REVIEW

File: pkg/didcomm/protocol/didexchange/service_test.go [hunk 1]
```diff
@@ -903,80 +939,77 @@ func TestService_Execute(t *testing.T) {
...
 	require.NoError(t, err)
 
 	// validate after unregister
-	require.Equal(t, 1, len(svc.statusEvents))
+	require.Equal(t, 1, len(svc.msgEvents))
+}
+
+func Test_AutoExecute(t *testing.T) {
+	ch := make(chan dispatcher.DIDCommEvent)
+	done := make(chan bool)
```
Comment (troyronda) [inline pkg/didcomm/protocol/didexchange/service_test.go]: FYI: done channel signatures are usually this:
`chan struct{}`
the bool is not needed.

You signal done with `close(chName)`.
You wait for done with `<-done`

Comment (rolsonquadras) [reply pkg/didcomm/protocol/didexchange/service_test.go]: [Updated](https://github.com/hyperledger/aries-framework-go/compare/eeed66994693b7c17d4b1c90ad61e09f2e22cbe7..4d12dff2d14ec87906c8c69146befde29694ebfe#diff-36ac39bebf32030fd7dff6f12aea8d7eR1006).

File: pkg/didcomm/protocol/didexchange/service.go [hunk 2]
```diff
@@ -124,90 +122,70 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
...
-// UnregisterMsg on DID Exchange protocol messages. Refer RegisterMsg().
-func (s *Service) UnregisterMsg(ch chan<- dispatcher.DIDCommMsg) error {
-	s.statusEventLock.Lock()
-	for i := 0; i < len(s.statusEvents); i++ {
-		if s.statusEvents[i] == ch {
-			s.statusEvents = append(s.statusEvents[:i], s.statusEvents[i+1:]...)
-			i--
-		}
+// UnregisterActionEvent on DID Exchange protocol messages. Refer RegisterActionEvent().
+func (s *Service) UnregisterActionEvent(ch chan<- dispatcher.DIDCommEvent) error {
```
Comment (troyronda) [inline pkg/didcomm/protocol/didexchange/service.go]: Why mismatch the channel type and the method name?

File: pkg/didcomm/protocol/didexchange/service.go [hunk 3]
```diff
@@ -124,90 +122,70 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
...
-// the workflow. This works in conjunction with RegisterEvent(). If this function is called after
-// RegisterEvent(), the service will execute the processing automatically and no action events will be triggered.
-// If both are not set, then service will not handle the messages and throw an error.
-func (s *Service) RegisterAutoExecute() error {
-	s.lock.Lock()
-	s.execute = true
-	s.lock.Unlock()
+// RegisterMsgEvent on DID Exchange protocol messages. The message events are triggered for incoming messages. Service
+// will not expect any callback on these events unlike Action events.
+func (s *Service) RegisterMsgEvent(ch chan<- dispatcher.DIDCommMsg) error {
```
Comment (troyronda) [inline pkg/didcomm/protocol/didexchange/service.go]: Similarly here. Why mismatch the channel type and the method name?

File: pkg/didcomm/protocol/didexchange/service.go [hunk 4]
```diff
@@ -432,3 +403,25 @@ func GenerateInviteWithKeyAndEndpoint(invite *Invitation) (string, error) {
 
 	return encodedExchangeInvitation(invite)
 }
+
+// AutoExecuteActionEvent is a utility function to execute events automatically. The function requires a channel to be
+// passed-in to listen for dispatcher.DIDCommEvent and triggers the callback. This is a blocking function and use
+// this function with a goroutine.
```
Comment (llorllale) [inline pkg/didcomm/protocol/didexchange/service.go]: Suggest including code snippet to show how auto execution is set up

File: pkg/didcomm/protocol/didexchange/service.go [hunk 5]
```diff
@@ -124,90 +122,70 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
...
-	s.lock.Lock()
-	s.actionEvent = nil
-	s.execute = false
-	s.lock.Unlock()
+// UnregisterMsgEvent on DID Exchange protocol messages. Refer RegisterMsgEvent().
+func (s *Service) UnregisterMsgEvent(ch chan<- dispatcher.DIDCommMsg) error {
+	s.msgEventLock.Lock()
+	for i := 0; i < len(s.msgEvents); i++ {
+		if s.msgEvents[i] == ch {
+			s.msgEvents = append(s.msgEvents[:i], s.msgEvents[i+1:]...)
```
Comment (llorllale) [inline pkg/didcomm/protocol/didexchange/service.go]: I think there's a panic here if `ch` is on the last index. Can you include a test for this scenario?

File: pkg/didcomm/protocol/didexchange/service.go (hunk 4 above)
Comment (rolsonquadras) [reply pkg/didcomm/protocol/didexchange/service.go]: [Added](https://github.com/hyperledger/aries-framework-go/compare/4d12dff2d14ec87906c8c69146befde29694ebfe..3526a9756b25f098cf4801865b8a7d1f506eb69b#diff-f761c91e753eba1c42a68076df06ce12R412).

File: pkg/didcomm/protocol/didexchange/service.go (hunk 2 above)
Comment (rolsonquadras) [reply pkg/didcomm/protocol/didexchange/service.go]: As discussed, [updated](https://github.com/hyperledger/aries-framework-go/compare/4d12dff2d14ec87906c8c69146befde29694ebfe..3526a9756b25f098cf4801865b8a7d1f506eb69b#diff-f761c91e753eba1c42a68076df06ce12R143) to match the format RegisterXYZEvent to DIDCommXYZ.

File: pkg/didcomm/protocol/didexchange/service.go (hunk 5 above)
Comment (rolsonquadras) [reply pkg/didcomm/protocol/didexchange/service.go]: Is [this](https://github.com/hyperledger/aries-framework-go/pull/321/files#diff-36ac39bebf32030fd7dff6f12aea8d7eR1003) fine ? Let me know if you see more tests required.

Comment (llorllale) [reply pkg/didcomm/protocol/didexchange/service.go]: Yeah, this feature of Golang is confusing: https://stackoverflow.com/questions/45299185/slice-start-position-greater-than-length-of-the-string

**Pull request merged**